src/katgpucbf/xbgpu/kernels/beamform.mako

```diff
@@ -0,0 +1,120 @@
+
+<%include file="/port.mako"/>
+
+#define N_BEAMS ${len(beam_pols)}
+#define BLOCK_TIME ${block_time}
+#define BLOCK_CHANNELS ${block_channels}
+#define QMAX 127
+
+// Quantise, and return saturation flag
+// TODO: this is copied from postproc.py. Create a shared implementation.
+DEVICE_FN bool quant(float value, int *out)
+{
+    float clamped = fminf(fmaxf(value, -QMAX), QMAX);
+    bool saturated = value != clamped;
+
+    int q;
+    // Convert to s8, round to nearest integer, and saturate
+    // (saturation is redundant due to the test above, but automatic
+    // for float-to-int conversions on CUDA hardware).
+#ifdef __OPENCL_VERSION__
+    q = convert_char_sat_rte(clamped);
+#else
+    asm("cvt.rni.sat.s8.f32 %0, %1;" : "=r" (q) : "f"(clamped));
+#endif
+    *out = q;
+    return saturated;
+}
+
+// Compute a * b in complex numbers
+DEVICE_FN float2 cmul(float2 a, float2 b)
+{
+    return make_float2(a.x * b.x - a.y * b.y, a.x * b.y + a.y * b.x);
+}
+
+// Compute a * b + c in complex numbers
+DEVICE_FN float2 cmad(float2 a, float2 b, float2 c)
+{
+    return make_float2(a.x * b.x - a.y * b.y + c.x, a.x * b.y + a.y * b.x + c.y);
+}
+
+// out: shape frame, beam, channel, time
+// in: shape frame, antenna, channel, time, pol
+// weight: shape antenna, beam, tightly packed
+// delays: shape antenna, beam, tightly packed
+// Each thread computes all beams for one (channel, time)
+KERNEL REQD_WORK_GROUP_SIZE(BLOCK_TIME, BLOCK_CHANNELS, 1) void beamform(
+    GLOBAL char2 * RESTRICT out,
+    GLOBAL const char4 * RESTRICT in,
+    GLOBAL const float2 * RESTRICT weights,
+    GLOBAL const float * RESTRICT delays,
+    int out_stride,
+    int out_beam_stride,
+    int out_frame_stride,
+    int in_stride,
+    int in_antenna_stride,
+    int in_frame_stride,
+    int n_antennas,
+    int n_channels,
+    int n_times
+)
+{
+    int time = get_global_id(0);
+    int channel = get_global_id(1);
+    int frame = get_group_id(2);
+    if (time >= n_times || channel >= n_channels)
+        return;  // out-of-bounds
+    in += frame * in_frame_stride + channel * in_stride + time;
+    out += frame * out_frame_stride + channel * out_stride + time;
+
+    float2 accum[N_BEAMS];
+    for (int i = 0; i < N_BEAMS; i++)
+        accum[i] = make_float2(0.0f, 0.0f);
+
+    for (int a = 0; a < n_antennas; a++)
+    {
+        char4 sample = *in;
+        float2 sample_pols[2] = {make_float2(sample.x, sample.y), make_float2(sample.z, sample.w)};
+        in += in_antenna_stride;
+% for i, pol in enumerate(beam_pols):
+        {
+            int i = ${i};
+            int pol = ${pol};
+
+            // TODO: w is time-invariant. Compute it once
+            // for all work-items with the same channel.
+            int addr = a * N_BEAMS + i;
+            float2 w = weights[addr];
+            float d = delays[addr] * channel;
+            float2 rot;
+            sincospif(d, &rot.y, &rot.x);
```

Comment:
TL;DR Would appreciate a general walkthrough of these few lines, even though they do seem simple/self-explanatory.

This might be me clutching at straws in attempting to understand these three, seemingly simple lines of code. I sought ~refuge in AvdB's [coeff_generator.py](https://github.com/ska-sa/dc_sand/blob/master/beamformer/beamforming/coeff_generator.py#L49-L68) in `dc_sand` for a bit of background/explanation. I don't pretend to understand the theory deeply, so the questions might be naive/unfounded
* Re: 'Phase correction' for band centre, is that something we need to take into account here?
* Or have we made an assumption upstream that addresses this? (e.g. In how delays are specified)

Perhaps you are just plainly doing things differently here, which is why it can be/is done so simply. Your thoughts, as always, much appreciated.

EDIT: Looking at the relevant dimensions between `delays` here and `delay_vals` in `dc_sand/beamformer`, I see they're both ultimately channel-dependent, but
* `Beamform.delays` - shape `(antennas, beams)` - here uses a `channel` value obtained via the kernel (no `channels_per_stream` dimension); whereas,
* `CoeffGenerator.delay_vals` - shape `(channels_per_stream, beams, antennas, 4 [?])` - uses both the delay and phase value to calculate the eventual phase rotation.

Completely understand if it's just better to focus on what's going on here, rather than trying to make comparisons between two separate (even entirely different) implementations.

Reply:
Take a look at the docstring for class `Beamform`: it documents that any constant term (phase_correction_band_center in Andrew's code, although there is also a user-provided phase term) is handled externally, by multiplying it into the weights (this is why the weights are complex rather than real).